security/src/test/java/org/kaazing/gateway/security/auth/BasicLoginModuleTest.java

```diff
@@ -135,17 +142,111 @@ public void testBasicLoginModuleIgnoresUnknownSchemes() throws Exception {
 
         context.assertIsSatisfied();
         assertNotNull(loginContext);
-        try {
-            loginContext.login();
-            final CallbackHandler nameCallbackHandler = handler.getDispatchMap().get(NameCallback.class);
-            final CallbackHandler passwordCallbackHandler = handler.getDispatchMap().get(PasswordCallback.class);
-            Assert.assertNull(nameCallbackHandler);
-            Assert.assertNull(passwordCallbackHandler);
-         } catch (LoginException e) {
-            fail("Login failed to succeed as expected: "+e.getMessage());
-        }
+
+        thrown.expect(LoginException.class);
+        loginContext.login();
+        final CallbackHandler nameCallbackHandler = handler.getDispatchMap().get(NameCallback.class);
+        final CallbackHandler passwordCallbackHandler = handler.getDispatchMap().get(PasswordCallback.class);
+        assertNull(nameCallbackHandler);
+        assertNull(passwordCallbackHandler);
+    }
+
+    @Test
+    public void testBasicLoginModuleFailsWhenNoLoginResultCallbackIsRegistered() throws Exception {
+        AuthenticationToken s = new DefaultAuthenticationToken("Basic", "Basic "+new String(Base64.encodeBase64("joe:welcome".getBytes())));
+        context.checking(new Expectations() {
+            {
+                oneOf(configuration).getAppConfigurationEntry(REALM_NAME);
+                final String loginModuleName = "org.kaazing.gateway.security.auth.BasicLoginModule";
+                final HashMap<String, Object> options = new HashMap<>();
+                final AppConfigurationEntry entry = new AppConfigurationEntry(loginModuleName,
+                        REQUIRED, options);
+                will(returnValue(new AppConfigurationEntry[]{entry}));
+            }
+        });
+
+        Subject subject = new Subject();
+        DispatchCallbackHandler handler = new DispatchCallbackHandler();
+        DefaultLoginResult loginResult = new DefaultLoginResult();
+        handler.register(AuthenticationTokenCallback.class, new AuthenticationTokenCallbackHandler(s));
+        LoginContext loginContext = new ResultAwareLoginContext("demo", subject, handler, configuration, loginResult);
+
+        context.assertIsSatisfied();
+        assertNotNull(loginContext);
+
+        thrown.expect(LoginException.class);
```

Comment:
There are several cases where we are doing exception handling on the callbacks, where the Gateway should always provide them.  I believe UnsupportedCallbackException would be thrown if the gateway failed to provide them and it was an error.  It seems to me, that the implementation of catching this exception and then throwing a LoginResultException with no reference to the missing callback handler is hiding the system/gateway error that it didn't provide that callback in the first place.

What happens if the original exception is allowed to propagate?(maybe internally on the JDK it get wrapped in a LoginResultException.  Would that be a better thing to have happen thereby showing it is a system error and not something implemented wrong in the login module?  Or perhaps there is some precedent for doing what we are doing, or a different solution somewhere in the JDK (maybe you can see what JDK classes extend Callback and where they are used).

This, combined with logging at the debug level in the HttpSecurityFilter the login exception (and other filters where we call this) I think may greatly decrease the amount of exception handling code we had to write.